packages/myst-migrate/src/v3_outputs.ts

```diff
@@ -0,0 +1,98 @@
+import { selectAll } from 'unist-util-select';
+import { assert } from 'console';
+import type { IFile } from './types.js';
+
+export const VERSION = 3;
+export const DATE = new Date(Date.parse('2025-04-09'));
+
+export const DESCRIPTION = `
+\`Output\` nodes previously could not represent AST trees for each output. 
+Now, the \`Outputs\` node has \`Output\` children with a 1:1 correspondence to \`IOutput\` bundles.
+
+Existing identifiers for the \`Output\` / \`Outputs\` nodes are not modified, as these are considered "content".
+`;
+
+type OutputV2 = {
+  type: 'output';
+  children?: any[];
+
+  data?: any[];
+  visibility?: any;
+
+  html_id?: string;
+  label?: string;
+  identifier?: string;
+
+  id?: string;
+};
+
+type OutputV3 = {
+  type: 'output';
+  children: any[];
+
+  jupyter_data?: any;
+
+  label?: string;
+  identifier?: string;
+  html_id?: string;
+};
+
+export function upgrade(file: IFile): IFile {
+  const { version, mdast } = file;
+  assert(version === VERSION, `Version must be ${VERSION}`);
+  const nodes = selectAll('output', mdast) as OutputV2[];
+  nodes.forEach((node) => {
+    // We can only correlate output children with the IOutput objects if there's only one IOutput
+    // Additionally, there may be placeholders that need to be removed
+    const numOutputs = node.data?.length ?? 0;
+    const children = numOutputs === 1 ? node.children ?? [] : [];
+    const placeholders = children.filter((child) => !!child.placeholder);
```

Comment:
The test case so far does not cover this logic (placeholders, multiple outputs, children preset), and should be extended to cover them.